subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -2827,6 +2827,118 @@ static int bt_hfp_ag_ccwa_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
 	return 0;
 }
 
+static void bt_hfp_ag_vr_activate(struct bt_hfp_ag *ag, void *user_data)
+{
+	bool feature;
+
+	hfp_ag_lock(ag);
+	feature = (ag->ag_features & BT_HFP_AG_FEATURE_ENH_VOICE_RECG) &&
+		(ag->hf_features & BT_HFP_HF_FEATURE_ENH_VOICE_RECG);
+	hfp_ag_unlock(ag);
+
+#if defined(CONFIG_BT_HFP_AG_VOICE_RECG)
+	if (bt_ag && bt_ag->voice_recognition) {
+		bt_ag->voice_recognition(ag, true);
+	} else {
+		(void)bt_hfp_ag_audio_connect(ag, BT_HFP_AG_CODEC_CVSD);
+	}
+#endif /* CONFIG_BT_HFP_AG_VOICE_RECG */
+
+	atomic_set_bit_to(ag->flags, BT_HFP_AG_VRE_R2A, !feature);
+
+#if defined(CONFIG_BT_HFP_AG_ENH_VOICE_RECG)
+	if (atomic_test_bit(ag->flags, BT_HFP_AG_VRE_R2A)) {
+		if (bt_ag && bt_ag->ready_to_accept_audio) {
+			bt_ag->ready_to_accept_audio(ag);
+		}
+	}
+#endif /* CONFIG_BT_HFP_AG_ENH_VOICE_RECG */
+}
+
+static void bt_hfp_ag_vr_deactivate(struct bt_hfp_ag *ag, void *user_data)
+{
+#if defined(CONFIG_BT_HFP_AG_ENH_VOICE_RECG)
+	if (bt_ag && bt_ag->voice_recognition) {
+		bt_ag->voice_recognition(ag, false);
+	}
+#endif /* CONFIG_BT_HFP_AG_ENH_VOICE_RECG */
+}
+
+static void bt_hfp_ag_vr_ready2accept(struct bt_hfp_ag *ag, void *user_data)
+{
+	atomic_set_bit(ag->flags, BT_HFP_AG_VRE_R2A);
+
+#if defined(CONFIG_BT_HFP_AG_ENH_VOICE_RECG)
+	if (bt_ag && bt_ag->ready_to_accept_audio) {
+		bt_ag->ready_to_accept_audio(ag);
+	}
+#endif /* CONFIG_BT_HFP_AG_ENH_VOICE_RECG */
+}
+
+static int bt_hfp_ag_bvra_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+{
+	int err;
+	uint32_t value;
+
+	hfp_ag_lock(ag);
+	if (!((ag->ag_features & BT_HFP_AG_FEATURE_VOICE_RECG) &&
+		  (ag->hf_features & BT_HFP_HF_FEATURE_VOICE_RECG))) {
```

Comment:
Incorrect split line alignment (there are *many* more places, but I won't point out each individual one). In general, I think it'd help make the code much more readable if you had a helper macro which takes the ag and a feature as input and returns a bool return value. That way you don't end up having both bitwise (`&`, `|`) and logical (`&&`, `||`) operators inside the same conditional expression.

Reply:
Updated